torch/functional.py

```diff
@@ -965,7 +965,7 @@ def tensordot(a, b, dims=2, out=None):
                 [ -0.2850,   4.2573,  -3.5997]])
     """
     if not torch.jit.is_scripting():
-        if (type(a) is not Tensor or type(b) is not Tensor) and has_torch_function((a, b)):
+        if has_torch_function((a, b)):
```

Comment:
This seems like a pessimization. Previously, tuple construction only happened if a was not a tensor and b was not a tensor. Now you unconditionally do tuple construction so you can call `has_torch_function`. I'm not too sure how CPython internally implements variadic function calls, but `has_torch_function(a, b)` has the potential to bypass this (although I guess if in the variadic case CPython constructs a tuple anyway, you'll need separate overloads per arity.)

Reply:
Surprisingly it's not. (kind of)

```
import re
from torch.utils.benchmark import Timer

for stmt in (
    "type(a) is not Tensor",
    "type(a) is not Tensor or type(b) is not Tensor",
    "object_has_torch_function(a)",
    "object_has_torch_function(a) or object_has_torch_function(b)",
    "has_torch_function((a, b))"
):
    timer = Timer(
        stmt,
        setup="""
        from torch import Tensor
        from torch.overrides import has_torch_function, object_has_torch_function
        a = torch.ones((1,))
        b = torch.ones((1,))
        # end_setup
        """
    )
    # Trim repeated `setup` for readability.
    s = repr(timer.blocked_autorange(min_run_time=2))
    print(re.sub(r"setup:.+# end_setup\n\n", "", s, flags=re.DOTALL), "\n")
```

```
<torch.utils.benchmark.utils.common.Measurement object at 0x7f254c3c5190>
type(a) is not Tensor
  Median: 73.92 ns
  IQR:    3.14 ns (72.39 to 75.52)
  268 measurements, 100000 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7f254c546c50>
type(a) is not Tensor or type(b) is not Tensor
  Median: 157.44 ns
  IQR:    6.42 ns (153.04 to 159.46)
  128 measurements, 100000 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7f254c3c5cd0>
object_has_torch_function(a)
  Median: 31.67 ns
  IQR:    1.40 ns (31.22 to 32.61)
  624 measurements, 100000 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7f2571ba7a50>
object_has_torch_function(a) or object_has_torch_function(b)
  Median: 60.68 ns
  IQR:    4.44 ns (58.10 to 62.55)
  323 measurements, 100000 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7f254c3c5bd0>
has_torch_function((a, b))
  Median: 65.73 ns
  IQR:    4.23 ns (63.92 to 68.16)
  296 measurements, 100000 runs per measurement, 1 thread
```

Even though the previous version wouldn't create a tuple for Tensor args, tuples are not that expensive and doing things in Python is slow. I checked, and while CPython has `METH_O`, there's no `METH_1` for static two arg functions so you'd have to go through `METH_VARARGS` and the tuple (or maybe `list`!!!) anyway. That said, one thing that I could do is to make a private version that statically expects a size two tuple as well to skip some of the unpacking work on the C++ side. (And MyPy will keep us honest on the calling side.) It won't be much, but since this is called everywhere it might be worth it to shave some nanoseconds.

Reply:
Ohh, actually `METH_FASTCALL` might be the thing.

Reply:
(It's 3.7+ though, so if it's useful it would have to be version guarded.)

Reply:
FASTCALL seems like the right thing, although `object_has_torch_function(a) or object_has_torch_function(b)` doesn't seem too shabby either. So while we're ripping up the floorboards right now, might as well design the API so that it is FASTCALL compatible.

Reply:
`FASTCALL` times:

```
<torch.utils.benchmark.utils.common.Measurement object at 0x7f5f511e3a90>
_has_torch_function_varargs(a)
  Median: 40.99 ns
  IQR:    1.53 ns (40.50 to 42.03)
  475 measurements, 100000 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7f5f75c649d0>
_has_torch_function_varargs(a, b)
  Median: 46.04 ns
  IQR:    1.33 ns (45.49 to 46.82)
  433 measurements, 100000 runs per measurement, 1 thread
```

So it's worth it to keep the single arg version, but for multiple it's considerably better to skip the intermediate tuple. And for Python 3.6 we can just use `METH_VARARGS` which isn't the end of the world. Particularly since we can statically use the tuple fast path that has to be checked in `THPModule_has_torch_function`.

Reply:
we're dumping py3.6 soon anyway

Reply:
By the way, I bet we have a lot of existing bindings that use `METH_VARARGS` and can switch to FASTCALL. So we should file a bug about this too.

Reply:
https://github.com/pytorch/pytorch/issues/49476